File: src/main/java/no/nav/sosialhjelp/soknad/business/db/repositories/soknadunderarbeid/SoknadUnderArbeidRowMapper.java
package no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import no.nav.sbl.soknadsosialhjelp.json.AdresseMixIn;
import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad;
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresse;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeidStatus;
import org.slf4j.Logger;
import org.springframework.jdbc.core.RowMapper;

import java.io.IOException;
import java.sql.ResultSet;
import java.sql.SQLException;

import static org.apache.commons.lang3.StringUtils.isNotEmpty;
import static org.slf4j.LoggerFactory.getLogger;

public class SoknadUnderArbeidRowMapper implements RowMapper<SoknadUnderArbeid>  {

    private static final Logger log = getLogger(SoknadUnderArbeidRowMapper.class);

    private final ObjectMapper mapper;
    private final ObjectWriter writer;

    {
        mapper = new ObjectMapper();
        mapper.addMixIn(JsonAdresse.class, AdresseMixIn.class);
        writer = mapper.writerWithDefaultPrettyPrinter();
    }

    public SoknadUnderArbeid mapRow(ResultSet rs, int rowNum) throws SQLException {
        SoknadUnderArbeidStatus status = null;
        try {
            final String statusFraDb = rs.getString("status");
            if (isNotEmpty(statusFraDb)) {
                status = SoknadUnderArbeidStatus.valueOf(statusFraDb);
            }
        } catch (IllegalArgumentException e) {
            throw new RuntimeException("Ukjent innsendingsstatus fra database", e);
        }
        return new SoknadUnderArbeid()
                .withSoknadId(rs.getLong("soknad_under_arbeid_id"))
                .withVersjon(rs.getLong("versjon"))
                .withBehandlingsId(rs.getString("behandlingsid"))
                .withTilknyttetBehandlingsId(rs.getString("tilknyttetbehandlingsid"))
                .withEier(rs.getString("eier"))
                .withJsonInternalSoknad(mapDataToJsonInternalSoknad(rs.getBytes("data")))
                .withStatus(status)
                .withOpprettetDato(rs.getTimestamp("opprettetdato") != null ?
                        rs.getTimestamp("opprettetdato").toLocalDateTime() : null)
                .withSistEndretDato(rs.getTimestamp("sistendretdato") != null ?
                        rs.getTimestamp("sistendretdato").toLocalDateTime() : null);
    }

    private JsonInternalSoknad mapDataToJsonInternalSoknad(byte[] data){
        if (data == null){
            return null;
        }
        try {
            return mapper.readValue(data, JsonInternalSoknad.class);
        } catch (IOException e) {
            log.error("Kunne ikke finne søknad", e);
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/db/repositories/soknadunderarbeid/SoknadUnderArbeidRowMapper.kt
package no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid

import no.nav.sbl.soknadsosialhjelp.json.JsonSosialhjelpObjectMapper
import no.nav.sbl.soknadsosialhjelp.soknad.JsonInternalSoknad
import org.slf4j.LoggerFactory
import org.springframework.jdbc.core.RowMapper
import java.io.IOException
import java.sql.ResultSet

class SoknadUnderArbeidRowMapper : RowMapper<SoknadUnderArbeid> {

    private val mapper = JsonSosialhjelpObjectMapper.createObjectMapper()

    override fun mapRow(rs: ResultSet, rowNum: Int): SoknadUnderArbeid {
        return SoknadUnderArbeid(
            soknadId = rs.getLong("soknad_under_arbeid_id"),
            versjon = rs.getLong("versjon"),
            behandlingsId = rs.getString("behandlingsid"),
            eier = rs.getString("eier"),
            jsonInternalSoknad = mapDataToJsonInternalSoknad(rs.getString("data")),
            status = SoknadUnderArbeidStatus.valueOf(rs.getString("status")),
            opprettetDato = rs.getTimestamp("opprettetdato").toLocalDateTime(),
            sistEndretDato = rs.getTimestamp("sistendretdato").toLocalDateTime()
        )
    }

    private fun mapDataToJsonInternalSoknad(data: String?): JsonInternalSoknad? {
        return data?.let {
            try {
                mapper.readValue(data, JsonInternalSoknad::class.java)
            } catch (e: IOException) {
                log.error("Kunne ikke finne søknad", e)
                throw RuntimeException(e)
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(SoknadUnderArbeidRowMapper::class.java)
    }
}
